keps/sig-storage/1495-volume-populators/README.md

```diff
@@ -271,6 +294,153 @@ downgrade.
+    - Components depending on the feature gate: kube-apiserver
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  Before this feature, kube API server would silently drop any PVC data source
+  that it didn't recognize, as if the client never populated the field. After
+  enabling this feature, the API server allows any object to be specified.
+  An external admission controller will take over validation of the field,
+  and its response to invalid data sources will be to fail the create operation,
+  rather than silently drop them.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Turning off the feature gate should drop all the data sources that were not
+  already valid and restore the old behavior.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  PVC which previously had data sources that were invalided by the rollback
+  would be empty volumes. Users would need to delete and re-create those PVCs.
+
+* **Are there any tests for feature enablement/disablement?**
+  Not at this time.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  As long as the validating webhook and its associated CRD is installed before
+  enabling the feature gate, the desired result will be achieved. Enabling the
+  feature gate before installing the validation hook will allow some PVCs to
+  be created with invalid data sources, and those PVCs will never get dynamically
+  provisioned.
+
+* **What specific metrics should inform a rollback?**
+  None.
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  Describe manual testing that was done and the outcomes.
+  Longer term, we may want to require automated upgrade/rollback tests, but we
+  are missing a bunch of machinery and tooling and can't do that now.
+
+* **Is the rollout accompanied by any deprecations and/or removals of features, APIs, 
+fields of API types, flags, etc.?**
+  No.
+
+### Monitoring Requirements
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can an operator determine if the feature is in use by workloads?**
+  Simply look at the data source field of the PVCs. Non-empty data sources with
+  a Kind other than snapshot and PVC indicate this feature is in use. Also the
+  existence of any VolumePopulator.populator.storage.k8s.io CRs would indicate
+  that a populator is installed and could be used.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine 
+the health of the service?**
+  - [ ] Metrics
```

Comment:
There are existing metrics implemented for volume provisioning, although we can't differentiate whether they are empty volumes or volumes with data sources.  Please list existing metrics for volume provisioning here.

Reply:
Do you have a link to those? I don't know anything about metrics.

Reply:
I have the following in the snapshot KEP.  It's the same for all sidecars:
```
csi_sidecar_operations_seconds (Container Storage Interface operation duration with gRPC error code status total. Reported from CSI external-snapshotter sidecar.)
```
I found the following but it is only for in-tree cloud providers.
```
https://kubernetes.io/docs/concepts/cluster-administration/system-metrics/#kube-controller-manager-metrics
```
@gnufied @msau42  are there any metrics for the PV controller that works for volumes provisioned by CSI drivers?

Reply:
I found this design doc: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-metrics.md
@gnufied does the metrics described in this design doc support volumes provisioned by CSI driver?

Reply:
metric names outstanding here.

Reply:
The in-tree metrics in PV controller do not work well for CSI because they assume provisioning is a synchronous operation, whereas with CSI it's async.

For CSI, we need to use the sidecar metrics for more accurate information

Reply:
This metric `csi_sidecar_operations_seconds` is for CSI sidecars.  Since this controller won't be implemented as a sidecar, this metric won't work here.

Reply:
I still don't see metrics here.  How about promoting some of the missing ones and getting them into the beta level?